From 5f607d0c2103c8d9512ab687b068b58bb4e26e31 Mon Sep 17 00:00:00 2001
From: Bart Ribbers <bribbers@disroot.org>
Date: Sun, 18 Apr 2021 00:13:42 +0200
Subject: [PATCH] Port to CMake

This project uses KDE's KCalendarCore. However, that package can only be
used with build systems that support CMake dependencies, which QMake is
not one of. KCalendarCore probably shipped a pkgconfig file in the past,
but this has not been the case for a quite a while as far as I can see.

Since using QMake thus prevents using a current KCalendarCore, switch
the build system to CMake instead
---
 .gitignore                                    |  1 +
 CMakeLists.txt                                | 71 +++++++++++++++++++
 doc/CMakeLists.txt                            |  7 ++
 doc/doc.pro                                   | 18 -----
 doc/libmkcal.cfg                              |  2 +-
 mkcal.pro                                     | 13 ----
 plugins/CMakeLists.txt                        |  1 +
 .../defaultinvitationplugin/CMakeLists.txt    | 15 ++++
 .../defaultinvitationplugin.pro               | 26 -------
 plugins/plugins.pro                           |  2 -
 src/CMakeLists.txt                            | 55 ++++++++++++++
 src/libmkcal-qt5.pc.in                        | 10 +++
 src/src.pro                                   | 61 ----------------
 tests/CMakeLists.txt                          | 19 +++++
 tests/tests.pro                               | 23 ------
 tools/CMakeLists.txt                          |  1 +
 tools/mkcaltool/CMakeLists.txt                | 16 +++++
 tools/mkcaltool/mkcaltool.pro                 | 21 ------
 tools/tools.pro                               |  2 -
 19 files changed, 197 insertions(+), 167 deletions(-)
 create mode 100644 CMakeLists.txt
 create mode 100644 doc/CMakeLists.txt
 delete mode 100644 doc/doc.pro
 delete mode 100644 mkcal.pro
 create mode 100644 plugins/CMakeLists.txt
 create mode 100644 plugins/defaultinvitationplugin/CMakeLists.txt
 delete mode 100644 plugins/defaultinvitationplugin/defaultinvitationplugin.pro
 delete mode 100644 plugins/plugins.pro
 create mode 100644 src/CMakeLists.txt
 create mode 100644 src/libmkcal-qt5.pc.in
 delete mode 100644 src/src.pro
 create mode 100644 tests/CMakeLists.txt
 delete mode 100644 tests/tests.pro
 create mode 100644 tools/CMakeLists.txt
 create mode 100644 tools/mkcaltool/CMakeLists.txt
 delete mode 100644 tools/mkcaltool/mkcaltool.pro
 delete mode 100644 tools/tools.pro

diff --git a/.gitignore b/.gitignore
index e8931eb..12fa738 100644
--- a/.gitignore
+++ b/.gitignore
@@ -11,3 +11,4 @@ doxygen.*
 *~
 *.so*
 plugins/defaultinvitationplugin/libdefaultinvitationplugin.so
+build
diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..721d324
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,71 @@
+cmake_minimum_required(VERSION 3.12.0)
+
+project(mkcal
+	VERSION 0.5.35
+	DESCRIPTION "Maemo mkcal calendar library")
+
+set(CMAKE_AUTOMOC ON)
+set(CMAKE_AUTORCC ON)
+set(CMAKE_AUTOUIC ON)
+
+set(CMAKE_INCLUDE_CURRENT_DIR ON)
+set(CMAKE_CXX_STANDARD 11)
+
+include(FindPkgConfig)
+include(FeatureSummary)
+include(GNUInstallDirs)
+include(CTest)
+
+set(INSTALL_QML_IMPORT_DIR "${CMAKE_INSTALL_FULL_LIBDIR}/qt/qml"
+	CACHE PATH "Custom QML import installation directory")
+option(BUILD_PLUGINS "Build plugin" ON)
+
+find_package(PkgConfig REQUIRED)
+find_package(ECM REQUIRED NO_MODULE)
+set(CMAKE_MODULE_PATH ${ECM_MODULE_PATH})
+
+set(QT_MIN_VERSION "5.6.0")
+find_package(Qt5 ${QT_MIN_VERSION} COMPONENTS DBus Gui Test REQUIRED)
+find_package(KF5 COMPONENTS CalendarCore REQUIRED)
+find_package(Doxygen)
+
+pkg_check_modules(TIMED timed-qt5 IMPORTED_TARGET REQUIRED)
+set_property(GLOBAL APPEND PROPERTY _CMAKE_timed-qt5_TYPE REQUIRED)
+pkg_check_modules(SQLITE3 sqlite3 IMPORTED_TARGET REQUIRED)
+set_property(GLOBAL APPEND PROPERTY _CMAKE_sqlite3_TYPE REQUIRED)
+
+if(TIMED_FOUND)
+	set_property(GLOBAL APPEND PROPERTY PACKAGES_FOUND timed-qt5)
+else()
+	set_property(GLOBAL APPEND PROPERTY PACKAGES_NOT_FOUND timed-qt5)
+endif()
+if(SQLITE3_FOUND)
+	set_property(GLOBAL APPEND PROPERTY PACKAGES_FOUND sqlite3)
+else()
+	set_property(GLOBAL APPEND PROPERTY PACKAGES_NOT_FOUND sqlite3)
+endif()
+
+add_subdirectory(src)
+add_subdirectory(tools)
+
+if(BUILD_PLUGINS)
+	pkg_check_modules(QMF QmfClient IMPORTED_TARGET REQUIRED)
+
+	if(QMF_FOUND)
+		set_property(GLOBAL APPEND PROPERTY PACKAGES_FOUND QmfClient)
+	else()
+		set_property(GLOBAL APPEND PROPERTY PACKAGES_NOT_FOUND QmfClient)
+	endif()
+
+	add_subdirectory(plugins)
+endif()
+
+if(BUILD_TESTING)
+	add_subdirectory(tests)
+endif()
+
+if(DOXYGEN_FOUND)	
+	add_subdirectory(doc)
+endif()
+
+feature_summary(WHAT ALL FATAL_ON_MISSING_REQUIRED_PACKAGES)
diff --git a/doc/CMakeLists.txt b/doc/CMakeLists.txt
new file mode 100644
index 0000000..778e473
--- /dev/null
+++ b/doc/CMakeLists.txt
@@ -0,0 +1,7 @@
+find_program(DOXYGEN NAMES doxygen)
+execute_process(
+	COMMAND ${DOXYGEN} ${CMAKE_CURRENT_SOURCE_DIR}/libmkcal.cfg; ${CMAKE_CURRENT_SOURCE_DIR}/doc/xmlize.pl;
+)
+
+install(DIRECTORY html
+	DESTINATION ${CMAKE_INSTALL_DATADIR}/doc/libmkcal-doc)
diff --git a/doc/doc.pro b/doc/doc.pro
deleted file mode 100644
index 2476803..0000000
--- a/doc/doc.pro
+++ /dev/null
@@ -1,18 +0,0 @@
-DOXYGEN_BIN=doxygen
-
-QMAKE_EXTRA_TARGETS += doc
-doc.target = doc
-isEmpty(DOXYGEN_BIN) {
-    doc.commands = @echo "Unable to detect doxygen in PATH"
-} else {
-    doc.commands = @$${DOXYGEN_BIN} doc/libmkcal.cfg ;
-    doc.commands+= ./doc/xmlize.pl ;
-}
-doc.depends = FORCE
-
-# Install rules
-htmldocs.files = ./doc/html/*
-htmldocs.path = /${DESTDIR}/usr/share/doc/libmkcal-doc
-htmldocs.CONFIG += no_check_exist
-
-INSTALLS += htmldocs    
diff --git a/doc/libmkcal.cfg b/doc/libmkcal.cfg
index 919ba9c..e5d49bc 100644
--- a/doc/libmkcal.cfg
+++ b/doc/libmkcal.cfg
@@ -69,7 +69,7 @@ GENERATE_HTML         = YES
 HTML_OUTPUT           = html
 HTML_HEADER           = doc/header.html
 HTML_FOOTER           = doc/footer.html
-HTML_STYLESHEET       = doc/libextendedkcal.css
+HTML_STYLESHEET       = doc/libmkcal.css
 HTML_ALIGN_MEMBERS    = YES
 GENERATE_HTMLHELP     = NO
 
diff --git a/mkcal.pro b/mkcal.pro
deleted file mode 100644
index 4842fd8..0000000
--- a/mkcal.pro
+++ /dev/null
@@ -1,13 +0,0 @@
-# QMake project file for libmkcal
-
-TEMPLATE = subdirs
-SUBDIRS = src tests tools
-
-SUBDIRS += plugins
-plugins.depends = src
-
-include(doc/doc.pro)
-
-CONFIG += ordered
-
-OTHER_FILES += rpm/*.spec
diff --git a/plugins/CMakeLists.txt b/plugins/CMakeLists.txt
new file mode 100644
index 0000000..3b859b6
--- /dev/null
+++ b/plugins/CMakeLists.txt
@@ -0,0 +1 @@
+add_subdirectory(defaultinvitationplugin)
diff --git a/plugins/defaultinvitationplugin/CMakeLists.txt b/plugins/defaultinvitationplugin/CMakeLists.txt
new file mode 100644
index 0000000..435c25a
--- /dev/null
+++ b/plugins/defaultinvitationplugin/CMakeLists.txt
@@ -0,0 +1,15 @@
+set(SRC
+	defaultinvitationplugin.cpp)
+set(HEADERS
+	defaultinvitationplugin.h)
+
+add_library(defaultinvitationplugin ${SRC} ${HEADERS})
+
+target_include_directories(defaultinvitationplugin PRIVATE ${PROJECT_SOURCE_DIR}/src)
+
+target_link_libraries(defaultinvitationplugin
+	KF5::CalendarCore
+	mkcal)
+
+install(TARGETS defaultinvitationplugin
+	DESTINATION ${CMAKE_INSTALL_LIBRARY_DIR}/mkcalplugins)
diff --git a/plugins/defaultinvitationplugin/defaultinvitationplugin.pro b/plugins/defaultinvitationplugin/defaultinvitationplugin.pro
deleted file mode 100644
index 145e9fb..0000000
--- a/plugins/defaultinvitationplugin/defaultinvitationplugin.pro
+++ /dev/null
@@ -1,26 +0,0 @@
-TEMPLATE = lib
-TARGET = $$qtLibraryTarget(defaultinvitationplugin)
-CONFIG += plugin
-OBJECTS_DIR = .obj
-MOC_DIR = .moc
-INCLUDEPATH +=  . \
-                ../.. \
-                ../../src \
-
-QMAKE_LIBDIR += ../../src
-QMAKE_CLEAN += *.so* .obj/* .moc/*
-
-LIBS += -L../../src/ -lmkcal-qt5
-CONFIG += link_pkgconfig
-PKGCONFIG += QmfClient \
-    KF5CalendarCore
-
-QT -= gui
-
-target.path +=  /${DESTDIR}$$[QT_INSTALL_LIBS]/mkcalplugins/
-
-INSTALLS += target
-
-HEADERS +=  defaultinvitationplugin.h
-
-SOURCES +=  defaultinvitationplugin.cpp
diff --git a/plugins/plugins.pro b/plugins/plugins.pro
deleted file mode 100644
index c0dfbae..0000000
--- a/plugins/plugins.pro
+++ /dev/null
@@ -1,2 +0,0 @@
-TEMPLATE = subdirs
-SUBDIRS = defaultinvitationplugin
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
new file mode 100644
index 0000000..bfbdad9
--- /dev/null
+++ b/src/CMakeLists.txt
@@ -0,0 +1,55 @@
+set(SRC
+	extendedcalendar.cpp
+	extendedstorage.cpp
+	notebook.cpp
+	sqliteformat.cpp
+	sqlitestorage.cpp
+	servicehandler.cpp
+	logging.cpp
+	semaphore_p.cpp)
+set(HEADERS
+	extendedcalendar.h
+	extendedstorage.h
+	extendedstorageobserver.h
+	notebook.h
+	sqliteformat.h
+	sqlitestorage.h
+	servicehandlerif.h
+	servicehandler.h
+	dummystorage.h
+	mkcal_export.h
+	logging_p.h
+	semaphore_p.h
+	invitationhandlerif.h
+	config-mkcal.h)
+
+add_library(mkcal-qt5 SHARED ${SRC} ${HEADERS})
+
+target_link_libraries(mkcal-qt5 PUBLIC
+	Qt5::DBus
+	Qt5::Gui
+	KF5::CalendarCore
+	PkgConfig::SQLITE3
+	PkgConfig::TIMED)
+
+set_target_properties(mkcal-qt5 PROPERTIES
+	SOVERSION ${PROJECT_VERSION_MAJOR}
+	VERSION ${PROJECT_VERSION})
+
+add_definitions(-fvisibility=hidden -fvisibility-inlines-hidden)
+add_definitions(-DMKCALPLUGINDIR="${CMAKE_INSTALL_LIBDIR}/mkcalplugins")
+
+# Install the library
+install(TARGETS mkcal-qt5
+	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})
+
+# Install headers
+install(FILES ${HEADERS}
+	DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/mkcal-qt5
+	COMPONENT Devel)
+
+configure_file(libmkcal-qt5.pc.in
+	${CMAKE_CURRENT_BINARY_DIR}/libmkcal-qt5.pc
+	@ONLY)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/libmkcal-qt5.pc
+	DESTINATION ${CMAKE_INSTALL_LIBDIR}/pkgconfig)
diff --git a/src/libmkcal-qt5.pc.in b/src/libmkcal-qt5.pc.in
new file mode 100644
index 0000000..5899f05
--- /dev/null
+++ b/src/libmkcal-qt5.pc.in
@@ -0,0 +1,10 @@
+prefix=@CMAKE_INSTALL_PREFIX@
+exec_prefix=${prefix}
+libdir=${prefix}/@CMAKE_INSTALL_LIBDIR@
+includedir=${prefix}/@CMAKE_INSTALL_INCLUDEDIR@/mkcal-qt5
+
+Name: mkcal-qt5
+Version: @PROJECT_VERSION@
+Description: @PROJECT_DESCRIPTION@
+Libs: -lmkcal-qt5
+Cflags: -I${includedir}
diff --git a/src/src.pro b/src/src.pro
deleted file mode 100644
index f3994a2..0000000
--- a/src/src.pro
+++ /dev/null
@@ -1,61 +0,0 @@
-# QMake project file for mkcal sources.
-TEMPLATE = lib
-TARGET = mkcal-qt5
-
-#DEFINES += MKCAL_FOR_MEEGO TIMED_SUPPORT
-
-DEFINES += MKCALPLUGINDIR=\\\"$$[QT_INSTALL_LIBS]/mkcalplugins\\\"
-
-DEFINES += TIMED_SUPPORT
-
-CONFIG += link_pkgconfig create_pc create_prl no_install_prl c++11
-PKGCONFIG += sqlite3 \
-    KF5CalendarCore \
-    timed-qt5
-
-QT += dbus
-QT -= gui
-
-target.path = $$[QT_INSTALL_LIBS]
-headers.path = /usr/include/mkcal-qt5
-headers.files += *.h \
-
-pkgconfig.path = $$[QT_INSTALL_LIBS]/pkgconfig
-
-INSTALLS += target \
-    headers \
-    pkgconfig
-
-QMAKE_CXXFLAGS += -fvisibility=hidden -fvisibility-inlines-hidden
-
-HEADERS += extendedcalendar.h \
-    extendedstorage.h \
-    notebook.h \
-    sqliteformat.h \
-    sqlitestorage.h \
-    servicehandlerif.h \
-    servicehandler.h \
-    dummystorage.h \
-    mkcal_export.h \
-    logging_p.h
-
-SOURCES += extendedcalendar.cpp \
-    extendedstorage.cpp \
-    notebook.cpp \
-    sqliteformat.cpp \
-    sqlitestorage.cpp \
-    servicehandler.cpp \
-    logging.cpp
-
-unix {
-    HEADERS += \
-        semaphore_p.h
-    SOURCES += \
-        semaphore_p.cpp
-}
-
-QMAKE_PKGCONFIG_NAME = $$TARGET
-QMAKE_PKGCONFIG_DESCRIPTION = Maemo mkcal calendar library
-QMAKE_PKGCONFIG_LIBDIR = $$target.path
-QMAKE_PKGCONFIG_INCDIR = $$headers.path
-QMAKE_PKGCONFIG_DESTDIR = pkgconfig
diff --git a/tests/CMakeLists.txt b/tests/CMakeLists.txt
new file mode 100644
index 0000000..fc6aaec
--- /dev/null
+++ b/tests/CMakeLists.txt
@@ -0,0 +1,19 @@
+set(SRC
+	tst_storage.cpp)
+set(HEADERS
+	tst_storage.h)
+
+add_executable(tst_storage ${SRC} ${HEADERS})
+
+target_include_directories(tst_storage PRIVATE ${PROJECT_SOURCE_DIR}/src)
+
+target_link_libraries(tst_storage
+	Qt5::DBus
+	Qt5::Test
+	KF5::CalendarCore
+	PkgConfig::SQLITE3
+	PkgConfig::TIMED
+	mkcal-qt5)
+
+enable_testing()
+add_test(tst_storage tst_storage)
diff --git a/tests/tests.pro b/tests/tests.pro
deleted file mode 100644
index 25f17a4..0000000
--- a/tests/tests.pro
+++ /dev/null
@@ -1,23 +0,0 @@
-TEMPLATE = app
-TARGET = tst_storage
-
-QT += testlib dbus
-CONFIG += link_pkgconfig c++11
-
-DEFINES += TIMED_SUPPORT
-
-INCLUDEPATH +=  ../src
-QMAKE_LIBDIR += ../src
-
-LIBS += -lmkcal-qt5
-PKGCONFIG += KF5CalendarCore sqlite3 timed-qt5
-
-HEADERS += \
-    tst_storage.h
-
-SOURCES += \
-    tst_storage.cpp
-
-target.path = /opt/tests/mkcal/
-
-INSTALLS += target
diff --git a/tools/CMakeLists.txt b/tools/CMakeLists.txt
new file mode 100644
index 0000000..0361272
--- /dev/null
+++ b/tools/CMakeLists.txt
@@ -0,0 +1 @@
+add_subdirectory(mkcaltool)
diff --git a/tools/mkcaltool/CMakeLists.txt b/tools/mkcaltool/CMakeLists.txt
new file mode 100644
index 0000000..0c165a4
--- /dev/null
+++ b/tools/mkcaltool/CMakeLists.txt
@@ -0,0 +1,16 @@
+set(SRC
+	main.cpp
+	mkcaltool.cpp)
+set(HEADERS
+	mkcaltool.h)
+
+add_executable(mkcaltool ${SRC} ${HEADERS})
+
+target_include_directories(mkcaltool PRIVATE ${PROJECT_SOURCE_DIR}/src)
+
+target_link_libraries(mkcaltool
+	KF5::CalendarCore
+	mkcal-qt5)
+
+install(TARGETS mkcaltool
+	DESTINATION ${CMAKE_INSTALL_BINDIR})
diff --git a/tools/mkcaltool/mkcaltool.pro b/tools/mkcaltool/mkcaltool.pro
deleted file mode 100644
index 31f47a0..0000000
--- a/tools/mkcaltool/mkcaltool.pro
+++ /dev/null
@@ -1,21 +0,0 @@
-TEMPLATE = app
-
-TARGET = mkcaltool
-target.path = /usr/bin
-
-CONFIG += link_pkgconfig
-
-INCLUDEPATH +=  ../../src
-QMAKE_LIBDIR += ../../src
-
-LIBS += -lmkcal-qt5
-PKGCONFIG += KF5CalendarCore
-
-
-SOURCES += main.cpp \
-    mkcaltool.cpp
-
-INSTALLS += target
-
-HEADERS += \
-    mkcaltool.h
diff --git a/tools/tools.pro b/tools/tools.pro
deleted file mode 100644
index 6317b61..0000000
--- a/tools/tools.pro
+++ /dev/null
@@ -1,2 +0,0 @@
-TEMPLATE = subdirs
-SUBDIRS =  mkcaltool
-- 
2.31.1

